File: src/test/java/io/protobase/f7/formulas/math/ATANTest.java
package io.protobase.f7.formulas.math;

import io.protobase.f7.errors.NAException;
import io.protobase.f7.errors.ValueException;
import io.protobase.f7.models.Grid;
import io.protobase.f7.testutils.TestFormula;
import org.junit.Test;

import static com.google.common.truth.Truth.assertThat;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

public class ATANTest extends TestFormula {
  @Test
  public void testApply() {
    assertThat(ATAN.SELF.apply(null, 10.0)).isEqualTo(Math.atan(10.0));
    assertThat(ATAN.SELF.apply(null, 128731.2)).isEqualTo(Math.atan(128731.2));
    assertThat(ATAN.SELF.apply(null, 11.11)).isEqualTo(Math.atan(11.11));
    assertThat(ATAN.SELF.apply(null, 0.0)).isEqualTo(Math.atan(0.0));
    assertThat(ATAN.SELF.apply(null, 88281.0)).isEqualTo(Math.atan(88281));
    assertThat(ATAN.SELF.apply(null, 2.0)).isEqualTo(Math.atan(2.0));
    assertThat(ATAN.SELF.apply(null, 4.0)).isEqualTo(Math.atan(4.0));
    assertThat(ATAN.SELF.apply(null, -4.0)).isEqualTo(Math.atan(-4.0));
    assertThat(ATAN.SELF.apply(null, -10124.0)).isEqualTo(Math.atan(-10124.0));
  }

  @Test
  public void testApply_stringConversion() {
    assertThat(ATAN.SELF.apply(null, "10.0")).isEqualTo(Math.atan(10.0));
  }

  @Test
  public void testApply_grid() {
    assertThat(ATAN.SELF.apply(null,
        Grid.builder().add(0, 0, -4.0).add(0, 1, "Don't mind me.").build()
    )).isEqualTo(Math.atan(-4.0));
  }

  @Test
  public void test_lookup() {
    ATAN F = new ATAN(lookup, collateralLookup);
    when(collateralLookup.apply(A1, M22_RANGE)).thenReturn(10.0);
    assertThat(F.apply(A1, M22_RANGE)).isEqualTo(Math.atan(10.0));
    verify(collateralLookup).apply(A1, M22_RANGE);
    verifyNoMoreInteractions(collateralLookup);
  }

  @Test
  public void testApply_errorsPassThrough() {
    assertThat(ATAN.SELF.apply(null, new ValueException())).isEqualTo(new ValueException());
  }

  @Test
  public void testApply_argumentsMismatch() {
    assertThat(ATAN.SELF.apply(null, "A", "Too many")).isEqualTo(new NAException());
  }
}

File: src/test/js/formulas/math/ATANTest.ts
import { assert } from "chai";
import { it, describe } from "../../testutils/TestUtils";
import { stub } from "sinon";
import { F7ExceptionName } from "../../../../main/js/errors/F7ExceptionName";
import { NAException } from "../../../../main/js/errors/NAException";
import { ValueException } from "../../../../main/js/errors/ValueException";
import { ATAN } from "../../../../main/js/formulas/math/ATAN";
import { Grid } from "../../../../main/js/models/common/Grid";
import { CollateralLookupFunction, LookupFunction } from "../../../../main/js/models/common/Types";
import { CommonModels } from "../../testutils/CommonModels";

describe("ATAN", function () {
  it("should work with valid numbers", function () {
    assert.deepEqual(ATAN.SELF.run(null, 10.0), Math.atan(10.0));
    assert.deepEqual(ATAN.SELF.run(null, 128731.2), Math.atan(128731.2));
    assert.deepEqual(ATAN.SELF.run(null, 11.11), Math.atan(11.11));
    assert.deepEqual(ATAN.SELF.run(null, 0.0), Math.atan(0.0));
    assert.deepEqual(ATAN.SELF.run(null, 88281.0), Math.atan(88281));
    assert.deepEqual(ATAN.SELF.run(null, 2.0), Math.atan(2.0));
    assert.deepEqual(ATAN.SELF.run(null, 4.0), Math.atan(4.0));
    assert.deepEqual(ATAN.SELF.run(null, -4.0), Math.atan(-4.0));
    assert.deepEqual(ATAN.SELF.run(null, -10124.0), Math.atan(-10124.0));
  });

  it("should work with strings", function () {
    assert.equal(ATAN.SELF.run(null, "1"), Math.atan(1));
  });

  it("should do pass-through errors", function () {
    assert.deepEqual(ATAN.SELF.run(null, new ValueException()), new ValueException());
  });

  it("should use lookup", function () {
    const lookup = stub();
    const collateralLookup = stub();
    const F = new ATAN(lookup as LookupFunction, collateralLookup as CollateralLookupFunction);
    collateralLookup.withArgs(CommonModels.A1, CommonModels.M22_RANGE).returns(0.99);
    assert.deepEqual(F.run(CommonModels.A1, CommonModels.M22_RANGE), Math.atan(0.99));
    assert.isTrue(collateralLookup.calledWith(CommonModels.A1, CommonModels.M22_RANGE));
    assert.equal(collateralLookup.callCount, 1);
    assert.isTrue(lookup.notCalled);
  });

  it("should handle grids", function () {
    const one = Grid.builder().add(0, 0, 0.99).add(0, 1, "A").build();
    assert.deepEqual(ATAN.SELF.run(null, one), Math.atan(0.99));
  });

  it("should return error when argument lengths are wrong", function () {
    assert.deepEqual((ATAN.SELF.run(null) as NAException).name, F7ExceptionName.NA);
    assert.deepEqual((ATAN.SELF.run(null, "A", "B") as NAException).name, F7ExceptionName.NA);
  });
});
